File: workers/ruby/spec/handlers/examples/blog_examples/post_04_team_scaling/step_handlers/payments/process_gateway_refund_handler.rb
# frozen_string_literal: true

module Payments
  module StepHandlers
    # TAS-137 Best Practices Demonstrated:
    # - get_input(): Access task context fields (refund_reason, partial_refund)
    # - get_input_or(): Access task context with defaults (refund_reason, partial_refund)
    # - get_dependency_result(): Access upstream step results (validate_payment_eligibility)
    # - get_dependency_field(): Extract nested fields from dependency results (payment_id, refund_amount, original_amount)
    class ProcessGatewayRefundHandler < TaskerCore::StepHandler::Base
      def call(context)
        # Extract and validate inputs
        inputs = extract_and_validate_inputs(context)

        logger.info "💰 ProcessGatewayRefundHandler: Processing gateway refund - task_uuid=#{context.task_uuid}, payment_id=#{inputs[:payment_id]}"

        # Simulate gateway refund processing
        refund_result = simulate_gateway_refund(inputs)

        # Ensure refund was successful
        ensure_refund_successful!(refund_result)

        logger.info "✅ ProcessGatewayRefundHandler: Refund processed - refund_id=#{refund_result[:refund_id]}, status=#{refund_result[:status]}"

        # Return standardized StepHandlerCallResult
        TaskerCore::Types::StepHandlerCallResult.success(
          result: {
            refund_processed: true,
            refund_id: refund_result[:refund_id],
            payment_id: inputs[:payment_id],
            refund_amount: inputs[:refund_amount],
            refund_status: refund_result[:status],
            gateway_transaction_id: refund_result[:gateway_transaction_id],
            gateway_provider: refund_result[:gateway_provider],
            processed_at: refund_result[:processed_at],
            estimated_arrival: refund_result[:estimated_arrival],
            namespace: 'payments'
          },
          metadata: {
            operation: 'process_gateway_refund',
            execution_hints: {
              refund_id: refund_result[:refund_id],
              gateway_provider: refund_result[:gateway_provider],
              refund_status: refund_result[:status]
            },
            http_headers: {
              'X-Gateway-Provider' => refund_result[:gateway_provider],
              'X-Refund-ID' => refund_result[:refund_id],
              'X-Gateway-Transaction-ID' => refund_result[:gateway_transaction_id]
            },
            input_refs: {
              refund_reason: 'context.get_input_or("refund_reason", "customer_request")',
              partial_refund: 'context.get_input_or("partial_refund", false)',
              payment_id: 'context.get_dependency_field("validate_payment_eligibility", "payment_id")',
              refund_amount: 'context.get_dependency_field("validate_payment_eligibility", "refund_amount")',
              original_amount: 'context.get_dependency_field("validate_payment_eligibility", "original_amount")'
            }
          }
        )
      rescue StandardError => e
        logger.error "❌ ProcessGatewayRefundHandler: Refund failed - #{e.class.name}: #{e.message}"
        raise
      end

      private

      # TAS-137: Extract and validate inputs using StepContext API
      def extract_and_validate_inputs(context)
        # TAS-137: Validate dependency result exists
        validation_result = context.get_dependency_result('validate_payment_eligibility')
        validation_result = validation_result.deep_symbolize_keys if validation_result

        unless validation_result&.dig(:payment_validated)
          raise TaskerCore::Errors::PermanentError.new(
            'Payment validation must be completed before processing refund',
            error_code: 'MISSING_VALIDATION'
          )
        end

        {
          # TAS-137: Use get_dependency_field for nested dependency access
          payment_id: context.get_dependency_field('validate_payment_eligibility', 'payment_id'),
          refund_amount: context.get_dependency_field('validate_payment_eligibility', 'refund_amount'),
          # TAS-137: Use get_input_or for task context with defaults
          refund_reason: context.get_input_or('refund_reason', 'customer_request'),
          partial_refund: context.get_input_or('partial_refund', false),
          # TAS-137: Use get_dependency_field for nested dependency access
          original_amount: context.get_dependency_field('validate_payment_eligibility', 'original_amount')
        }
      end

      # Simulate gateway refund processing (self-contained)
      def simulate_gateway_refund(inputs)
        payment_id = inputs[:payment_id]

        # Simulate different gateway responses
        case payment_id
        when /pay_test_gateway_timeout/
          { status: 'timeout', error: 'Gateway timeout' }
        when /pay_test_gateway_error/
          { status: 'failed', error: 'Gateway error' }
        when /pay_test_rate_limit/
          { status: 'rate_limited', error: 'Rate limit exceeded' }
        else
          # Success case
          refund_id = "rfnd_#{SecureRandom.hex(12)}"
          {
            status: 'processed',
            refund_id: refund_id,
            payment_id: payment_id,
            gateway_transaction_id: "gtx_#{SecureRandom.hex(10)}",
            gateway_provider: 'MockPaymentGateway',
            refund_amount: inputs[:refund_amount],
            processed_at: Time.now.utc.iso8601,
            estimated_arrival: (Time.now + 5.days).utc.iso8601
          }
        end
      end

      # Ensure refund was processed successfully
      def ensure_refund_successful!(refund_result)
        status = refund_result[:status]

        case status
        when 'processed', 'succeeded'
          # Refund successful
          nil
        when 'failed'
          # Permanent error - refund failed
          raise TaskerCore::Errors::PermanentError.new(
            "Gateway refund failed: #{refund_result[:error]}",
            error_code: 'GATEWAY_REFUND_FAILED'
          )
        when 'timeout'
          # Temporary error - gateway timeout
          raise TaskerCore::Errors::RetryableError.new(
            'Gateway timeout, will retry',
            retry_after: 30
          )
        when 'rate_limited'
          # Temporary error - rate limit
          raise TaskerCore::Errors::RetryableError.new(
            'Gateway rate limited, will retry',
            retry_after: 60
          )
        when 'pending'
          # Temporary state - refund pending
          raise TaskerCore::Errors::RetryableError.new(
            'Refund pending, checking again',
            retry_after: 15
          )
        else
          # Unknown status - treat as retryable
          raise TaskerCore::Errors::RetryableError.new(
            "Unknown refund status: #{status}",
            retry_after: 30
          )
        end
      end
    end
  end
end

File: workers/python/tests/handlers/examples/blog_examples/post_04_team_scaling/payments/process_gateway_refund_handler.py
"""
Process Gateway Refund Handler for Blog Post 04 - Payments namespace.

Executes the refund through the payment processor gateway.

TAS-137 Best Practices:
- get_input(): Access task context fields (refund_reason, partial_refund)
- get_input_or(): Access task context with defaults (refund_reason, partial_refund)
- get_dependency_result(): Access upstream step results (validate_payment_eligibility)
- get_dependency_field(): Extract nested fields (payment_id, refund_amount, original_amount)
"""

import uuid
from datetime import datetime, timedelta, timezone

from tasker_core import StepContext, StepHandler, StepHandlerResult


class ProcessGatewayRefundHandler(StepHandler):
    """Process refund through payment gateway."""

    handler_name = "team_scaling.payments.step_handlers.ProcessGatewayRefundHandler"
    handler_version = "1.0.0"

    def call(self, context: StepContext) -> StepHandlerResult:
        """Execute the process gateway refund step."""
        # TAS-137: Validate dependency result exists
        validation_result = context.get_dependency_result("validate_payment_eligibility")

        if not validation_result or not validation_result.get("payment_validated"):
            return self.failure(
                "Payment validation must be completed before processing refund",
                error_code="MISSING_VALIDATION",
                retryable=False,
            )

        # TAS-137: Use get_dependency_field() for nested extraction
        payment_id = context.get_dependency_field("validate_payment_eligibility", "payment_id")
        refund_amount = context.get_dependency_field(
            "validate_payment_eligibility", "refund_amount"
        )
        _original_amount = context.get_dependency_field(
            "validate_payment_eligibility", "original_amount"
        )  # For audit logging

        # TAS-137: Use get_input_or() for task context with defaults
        refund_reason = context.get_input_or("refund_reason", "customer_request")
        _partial_refund = context.get_input_or("partial_refund", False)  # For enhanced processing

        # Simulate gateway refund processing
        refund_result = self._simulate_gateway_refund(payment_id, refund_amount, refund_reason)

        # Ensure refund was successful
        refund_error = self._ensure_refund_successful(refund_result)
        if refund_error:
            return refund_error

        return self.success(
            {
                "refund_processed": True,
                "refund_id": refund_result["refund_id"],
                "payment_id": payment_id,
                "refund_amount": refund_amount,
                "refund_status": refund_result["status"],
                "gateway_transaction_id": refund_result["gateway_transaction_id"],
                "gateway_provider": refund_result["gateway_provider"],
                "processed_at": refund_result["processed_at"],
                "estimated_arrival": refund_result["estimated_arrival"],
                "namespace": "payments",
            },
            {
                "operation": "process_gateway_refund",
                "service": "payment_gateway",
                "refund_id": refund_result["refund_id"],
                "gateway_provider": refund_result["gateway_provider"],
            },
        )

    def _simulate_gateway_refund(
        self, payment_id: str, refund_amount: float, _refund_reason: str
    ) -> dict:
        """Simulate gateway refund processing."""
        now = datetime.now(timezone.utc)

        # Simulate different gateway responses based on payment_id patterns
        if "pay_test_gateway_timeout" in payment_id:
            return {"status": "timeout", "error": "Gateway timeout"}
        elif "pay_test_gateway_error" in payment_id:
            return {"status": "failed", "error": "Gateway error"}
        elif "pay_test_rate_limit" in payment_id:
            return {"status": "rate_limited", "error": "Rate limit exceeded"}
        else:
            # Success case
            return {
                "status": "processed",
                "refund_id": f"rfnd_{uuid.uuid4().hex[:24]}",
                "payment_id": payment_id,
                "gateway_transaction_id": f"gtx_{uuid.uuid4().hex[:20]}",
                "gateway_provider": "MockPaymentGateway",
                "refund_amount": refund_amount,
                "processed_at": now.isoformat(),
                "estimated_arrival": (now + timedelta(days=5)).isoformat(),
            }

    def _ensure_refund_successful(self, refund_result: dict) -> StepHandlerResult | None:
        """Ensure refund was processed successfully."""
        status = refund_result.get("status")

        if status in ("processed", "succeeded"):
            return None  # Success
        elif status == "failed":
            return self.failure(
                f"Gateway refund failed: {refund_result.get('error')}",
                error_code="GATEWAY_REFUND_FAILED",
                retryable=False,
            )
        elif status == "timeout":
            return self.failure(
                "Gateway timeout, will retry",
                error_code="GATEWAY_TIMEOUT",
                retryable=True,
            )
        elif status == "rate_limited":
            return self.failure(
                "Gateway rate limited, will retry",
                error_code="RATE_LIMITED",
                retryable=True,
            )
        elif status == "pending":
            return self.failure(
                "Refund pending, checking again",
                error_code="REFUND_PENDING",
                retryable=True,
            )
        else:
            return self.failure(
                f"Unknown refund status: {status}",
                error_code="UNKNOWN_STATUS",
                retryable=True,
            )
